1.segment-data/cell-health-segmentation.py

```diff
@@ -0,0 +1,223 @@
+    )
+    # remove cell masks if they are on the edge
+    if model_specs["remove_edge_masks"]:
+        masks = utils.remove_edge_masks(masks)
+    
+    outlines = utils.outlines_list(masks)
+    for outline in outlines:
+        centroid = outline.mean(axis=0)
+        object_data = {
+            "Location_Center_X": centroid[0],
+            "Location_Center_Y": centroid[1],
+        }
+        objects_data.append(object_data)
+
+    objects_data = pd.DataFrame(objects_data)
+    return objects_data
+
+
+def overlay_channels(identifier: str, current_dir: pathlib.Path) -> np.ndarray:
+    """overlays nuclei, ER, and RNA channels to help with CellPose segmentation
+
+    Args:
+        identifier (str): string of current field to overlay channels for
+        current_dir (pathlib.Path): directory of current field
+
+    Returns:
+        np.ndarray: 3 channel image with nuclei, ER, RNA
+    """
+    # load all channel images
+    channel_paths = []
+    channel_images = []
+    for image_file in current_dir.iterdir():
+        if identifier in image_file.name:
+            channel_paths.append(image_file)
+    channel_paths = sorted(channel_paths)
+    for channel_path in channel_paths:
+        channel_image = np.array(cv2.imread(str(channel_path), 0))
+        channel_images.append(channel_image)
+
+    # channel multipliers are only useful for human interpretation of the overlayed images
+    # channel brightness is irrelevant to CellPose segmentation
+    # channel images 1,2,3 are DNA, ER, RNA respectively
+    overlay = np.dstack(
+        [channel_images[0] * 2, channel_images[1] * 1, channel_images[2] * 5]
+    ).astype(np.uint8)
+
+    return overlay
+
+
+def get_nuc_cyto_data(nuc_locations: pd.DataFrame, cyto_outlines: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """produces nuclei and cytoplasm dataframes with corresponding cell IDs from nuclei locations and cytoplasm outlines
+
+    Args:
+        nuc_locations (pd.DataFrame): dataframe with nuclei center coords for certain image
+        cyto_outlines (pd.DataFrame): dataframe with cytoplasm outline for same image as above
+
+    Returns:
+        Tuple[pd.DataFrame, pd.DataFrame]: nuc_data and cyto_data with corresponding cell IDs and center coords for each object
+    """
+    nuc_data = []
+    cyto_data = []
+    
+    # add Cell_ID column, each ID corresponds to index of outline entry
+    cyto_outlines = cyto_outlines.reset_index()
+    cyto_outlines = cyto_outlines.rename(columns={'index':'Cell_ID'})
+    
+    # iterate over all cytoplasm outlines
+    for cyto_index, cyto_row in cyto_outlines.iterrows():
+        cell_id = cyto_row["Cell_ID"]
+        
+        # create path (polygon) from outline to check if the nuclei are in this path
+        outline = cyto_row["Outline"]
+        cytoplasm_path = mplPath.Path(outline)
+        # check if any of the nuclei are within cytoplasm path
+        for nuc_index, nuc_row in nuc_locations.iterrows():
```

Comment:
this looks like a time sink to me!

It's probably premature to optimize now (a general rule of thumb is to leave optimization for later, after a baseline implementation).

In the future, if we do need to optimize, this is a good candidate place to start.

Reply:
Slightly optimized with code below. More changes to do in the future.

```
# drop nucleus so it doesn't have to be rechecked
nuc_locations.drop(nuc_index, inplace=True)
```